pkg/api/state_test.go

```diff
@@ -0,0 +1,225 @@
+	"fmt"
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/assert"
+	"log"
+	"strconv"
+	"testing"
+)
+
+func TestExample(t *testing.T) {
+	b, err := json.Marshal(map[string]*State{
+		"foo": {
+			Role: "goalkeeper",
+		},
+		"bar": {
+			Role: "goalsaver",
+		},
+	})
+	if err != nil {
+		log.Fatal(err)
+	}
+	fmt.Println(string(b))
+}
+
+func TestGeneral(t *testing.T) {
+	activeCase := getNewValidState()
+	validate(t, activeCase, true, "test 1")
+
+	activeCase.VisionStatus = "Manual" // capital
+	validate(t, activeCase, false, "test 2 capital")
+
+	activeCase = getNewValidState()
+	activeCase.Temperature1 = 28
+	activeCase.CamStatus = 12 // too much
+	activeCase.Cpu1Load = 89
+	activeCase.Cpu0Load = 87
+	activeCase.EmergencyStatus = 0
+	validate(t, activeCase, false, "test 3 pass upperbound")
+
+	activeCase = getNewValidState()
+	activeCase.EmergencyStatus = 100 // maximum
+	validate(t, activeCase, true, "test 4 maximum")
+
+	activeCase = getNewValidState()
+	activeCase.BatteryVoltage = 255 // max_uint8
+	validate(t, activeCase, false, "test 5 max_uint8")
+
+	activeCase = getNewValidState()
+	activeCase.ID = "" // empty ID
+	validate(t, activeCase, false, "test 6 empty ID")
+}
+
+func validate(t *testing.T, state State, shouldWork bool, testName string) {
+	err := state.Validate()
+	if shouldWork {
+		assert.Nil(t, err, testName)
+
+	} else {
+		assert.NotNil(t, err, testName)
+	}
+}
+
+// checks every field of the state, returning an error when any invalid field is encountered
+func (state State) Validate() error {
+
+	if state.ID == "" {
+		return errors.New("Empty ID header")
+	}
+	// first evaluate all tests, then return the first error.
+	// probably the most space-efficient method of mass-checking
+	results := []error{
+		state.VisionStatus.Validate(),
+		state.BallFound.Validate(),
+		state.LocalizationStatus.Validate(),
+		state.CPB.Validate(),
+		state.Role.Validate(),
+		state.HomeGoal.Validate(),
+		state.TeamColor.Validate(),
+		state.CapacitorState.Validate(),
+		state.Kinect1State.Validate(),
+		check("Restart counter of Motion Executable", int(state.RestartCountMotion), 0, 99),
+		check("Restart counter of Vision Executable", int(state.RestartCountVision), 0, 99),
+		check("Restart counter of World Model", int(state.RestartCountWorldmodel), 0, 99),
+		check("Battery voltage", int(state.BatteryVoltage), 0, 99),
+		check("Emergency status", int(state.EmergencyStatus), 0, 100),
+		check("Cpu 0 load", int(state.Cpu0Load), 0, 99),
+		check("Cpu 1 load", int(state.Cpu1Load), 0, 99),
+		check("Active dev pc", int(state.ActiveDevPC), 0, 90),
+		check("Temperature 1", int(state.Temperature1), 0, 99),
+		check("Temperature 2", int(state.Temperature2), 0, 99),
+		check("Temperature 3", int(state.Temperature3), 0, 99),
+		check("Cam status", int(state.CamStatus), 0, 10),
+	}
+
+	for _, e := range results {
+		if e != nil {
+			return e
+		}
+	}
+
+	return nil
+}
+
+func check(name string, value int, min int, max int) error {
```

Comment:
You don't need this function. Better define a default error constructor/format for that type of errors.